python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_execution_context.py

```diff
@@ -0,0 +1,186 @@
+        "resources",
+        "run_tags",
+        "selected_asset_check_keys",
+        "selected_asset_keys",
+        "selected_output_names",
+        "set_data_version",
+        "set_requires_typed_event_stream",
+        "typed_event_stream_error_message",
+        "describe_op",
+        "file_manager",
+        "has_assets_def",
+        "get_mapping_key",
+        "get_step_execution_context",
+        "job_def",
+        "node_handle",
+        "op",
+        "op_config",
+        "op_def",
+        "op_handle",
+        "step_launcher",
+        "has_events",
+        "consume_events",
+        "log_event",
+        "get_asset_provenance",
+        "is_subset",
+        "partition_keys",
+        "get",
+    ]
+
+    other_ignores = [
+        "_abc_impl",
+        "_events",
+        "_output_metadata",
+        "_pdb",
+        "_step_execution_context",
+    ]
+
+    def assert_deprecation_messages_as_expected(received_info, expected_info):
+        assert received_info.breaking_version == expected_info["breaking_version"]
+        assert received_info.additional_warn_text == expected_info["additional_warn_text"]
+        assert received_info.subject == expected_info["subject"]
+
+    @asset
+    def test_context(context: AssetExecutionContext):
+        asset_context = context
+        op_context = context.op_execution_context
+
+        op_context_properties = []
+
+        for attr in dir(OpExecutionContext):
+            if isinstance(getattr(OpExecutionContext, attr), property):
+                op_context_properties.append(attr)
+                if attr in asset_context_not_deprecated:
+                    try:
+                        getattr(AssetExecutionContext, attr)
+                    except Exception:
+                        raise Exception(
+                            f"Property {attr} on OpExecutionContext does not have an implementation on"
+                            " AssetExecutionContext. All methods on OpExecutionContext must be"
+                            " re-implemented on AssetExecutionContext See the class implementation of"
+                            " AssetExecutionContext for more details."
+                        )
+                else:
+                    try:
+                        deprecation_info = getattr(  # noqa: SLF001
+                            AssetExecutionContext, attr
+                        ).fget._deprecated
+
+                    except Exception:
+                        raise Exception(
+                            f"Property {attr} on OpExecutionContext does not have an implementation on"
+                            " AssetExecutionContext. All methods on OpExecutionContext must be"
+                            " re-implemented on AssetExecutionContext with appropriate deprecation"
+                            " warnings. See the class implementation of AssetExecutionContext for more"
+                            " details."
+                        )
```

Comment:
Seems like both of these try/except blocks are just testing if `getattr` throws an error. IMO would be better to just `if not hasattr(...)` and throw the exception instead of catching generic `Exception`

Reply:
The two error messages are very similar but contradictory:

- "All methods ... must be re-implemented ... with appropriate deprecation warnings."
- "All methods ... must be re-implemented"

Would recommend making it a single error message and caveating inside with need for deprecation warnings if not in the "not deprecated" list.

Reply:
Also should use "All properties" instead of "All methods" here